labgrid/driver/rawnetworkinterfacedriver.py

```diff
@@ -0,0 +1,130 @@
+# pylint: disable=no-member
+import contextlib
+import json
+import signal
+import subprocess
+
+import attr
+
+from .common import Driver
+from ..factory import target_factory
+from ..step import step
+from ..util.helper import processwrapper
+from ..util.managedfile import ManagedFile
+from ..resource.common import NetworkResource
+
+
+@target_factory.reg_driver
+@attr.s(eq=False)
+class RawNetworkInterfaceDriver(Driver):
+    bindings = {
+        "iface": {"NetworkInterface", "RemoteNetworkInterface", "USBNetworkInterface"},
+    }
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        self._record_handle = None
+        self._replay_handle = None
+
+    def _wrap_command(self, args):
+        wrapper = ["sudo", "labgrid-raw-interface"]
+
+        if self.iface.command_prefix:
+            # add ssh prefix, convert command passed via ssh (including wrapper) to single argument
+            return self.iface.command_prefix + [" ".join(wrapper + args)]
+        else:
+            # keep wrapper and args as-is
+            return wrapper + args
+
+    def _stop(self, proc, *, timeout=None):
+        assert proc is not None
+
+        try:
+            _, err = proc.communicate(timeout=timeout)
+        except subprocess.TimeoutExpired:
+            proc.send_signal(signal.SIGINT)
+            _, err = proc.communicate()
+            raise
+
+        if proc.returncode:
+            raise subprocess.CalledProcessError(
+                returncode=proc.returncode,
+                cmd=proc.args,
+                stderr=err,
+            )
+
+    @Driver.check_active
+    @step(args=["filename", "count"])
+    def start_record(self, filename, *, count=None):
+        assert self._record_handle is None
+
+        cmd = ["tcpdump", self.iface.ifname]
+        if count is not None:
+            cmd.append(str(count))
+        cmd = self._wrap_command(cmd)
+        outdata = open(filename, "wb")
+        self._record_handle = subprocess.Popen(cmd, stdout=outdata, stderr=subprocess.PIPE)
+        return self._record_handle
+
+    @Driver.check_active
+    @step(args=["timeout"])
+    def stop_record(self, *, timeout=None):
+        try:
+            self._stop(self._record_handle, timeout=timeout)
+        finally:
+            self._record_handle = None
+
+    @contextlib.contextmanager
+    def record(self, filename, *, count=None, timeout=None):
```

Comment:
Do we need to ensure that one of `count` or `timeout` is set? Otherwise tcpdump won't ever stop...

Perhaps some docstrings for these functions would be useful. I think, in most cases a timeout should be specified, as it avoids the possibility of just having a stuck test.

Reply:
> Do we need to ensure that one of `count` or `timeout` is set? Otherwise tcpdump won't ever stop...

I've added an assert to check this.

> Perhaps some docstrings for these functions would be useful. I think, in most cases a timeout should be specified, as it avoids the possibility of just having a stuck test.

Docstrings added.